### DISCUSSION TITLE
[Forge 0.0.16] NeverOOM Checkbox

### DISCUSSION THREAD

**Discussion by lllyasviel** [lllyasviel/stable-diffusion-webui-forge#395] [Announcements]
![image](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/19834515/b269bfdb-513f-49cb-b298-1a7acf5f483e)

These are some two QoL tricks for making large images.

If you use `Enabled for VAE (always tiled)` you will always use tiled VAE to encode/decode images.

If you use `Enabled for UNet (always maximize offload)`, the diffusion GPU memory will drop to smaller than 1.5GB for SDXL at 1024x1024 (and even smaller for SD1.5). For example, this is me generating 1024x1024 image using SDXL base model with `Enabled for UNet` checked

![image](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/19834515/a1a5af2c-13bd-4fa1-a674-076ccf2c928f)
![image](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/19834515/8fc1575d-4fcf-4c8b-a120-aa2eca6b2fa1)

![image](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/19834515/b248979c-d6d2-447e-bd43-fdefb43bf67d)

You can see that this GPU memory is always less than 1.5GB even for SDXL at 1024px.

Using this you can make use of all your remaining memory to get super big images.

By using `Enabled for UNet (always maximize offload)`, I am able to generate images at 6553x6553 using 8GB vram with SDXL. Previously this would need multi-diffusion to sacrifice result quality a bit but now you can generate at that resolution natively in one single pass. 

Below is screenshots with 8GB vram diffuse at 6553x6553 natively using SDXL

![image](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/19834515/ef4134e0-4798-4d52-8bf3-24453130f057)

![image](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/19834515/de98243d-b787-4803-8b49-2bee23328e64)

**Comment by lllyasviel** [comment]
Besides recently we have also added a few cmd flags to tune the performance.

The below texts are taken from Readme:

1. `--always-offload-from-vram` (This flag will make things **slower** but less risky). This option will let Forge always unload models from VRAM. This can be useful if you use multiple software together and want Forge to use less VRAM and give some VRAM to other software, or when you are using some old extensions that will compete vram with Forge, or (very rarely) when you get OOM.

2. `--cuda-malloc` (This flag will make things **faster** but more risky). This will ask pytorch to use *cudaMallocAsync* for tensor malloc. On some profilers I can observe performance gain at millisecond level, but the real speed up on most my devices are often unnoticed (about or less than 0.1 second per image). This cannot be set as default because many users reported issues that the async malloc will crash the program. Users need to enable this cmd flag at their own risk.

3. `--cuda-stream` (This flag will make things **faster** but more risky). This will use pytorch CUDA streams (a special type of thread on GPU) to move models and compute tensors simultaneously. This can almost eliminate all model moving time, and speed up SDXL on 30XX/40XX devices with small VRAM (eg, RTX 4050 6GB, RTX 3060 Laptop 6GB, etc) by about 15\% to 25\%. However, this unfortunately cannot be set as default because I observe higher possibility of pure black images (Nan outputs) on 2060, and higher chance of OOM on 1080 and 2060. When the resolution is large, there is a chance that the computation time of one single attention layer is longer than the time for moving entire model to GPU. When that happens, the next attention layer will OOM since the GPU is filled with the entire model, and no remaining space is available for computing another attention layer. Most overhead detecting methods are not robust enough to be reliable on old devices (in my tests). Users need to enable this cmd flag at their own risk.

4. `--pin-shared-memory` (This flag will make things **faster** but more risky). Effective only when used together with `--cuda-stream`. This will offload modules to Shared GPU Memory instead of system RAM when offloading models. On some 30XX/40XX devices with small VRAM (eg, RTX 4050 6GB, RTX 3060 Laptop 6GB, etc), I can observe significant (at least 20\%) speed-up for SDXL. However, this unfortunately cannot be set as default because the OOM of Shared GPU Memory is a much more severe problem than common GPU memory OOM. Pytorch does not provide any robust method to unload or detect Shared GPU Memory. Once the Shared GPU Memory OOM, the entire program will crash (observed with SDXL on GTX 1060/1050/1066), and there is no dynamic method to prevent or recover from the crash. Users need to enable this cmd flag at their own risk.

**Reply by ositoMalvado** [reply]
> 3\. -

@lllyasviel hi master, I want to know how to use Never OOM options by calling API, any way?

**Reply by TimmekHW** [reply]
I support you! I also want to change the status or save Never OOM with API
![image](https://github.com/user-attachments/assets/2fafa800-5d8a-46df-8fb4-99b44b64e203)

**Reply by ositoMalvado** [reply]
> I support you! I also want to change the status or save Never OOM with API
> ![image](https://private-user-images.githubusercontent.com/94626112/361957630-2fafa800-5d8a-46df-8fb4-99b44b64e203.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkzODIzMDIsIm5iZiI6MTcyOTM4MjAwMiwicGF0aCI6Ii85NDYyNjExMi8zNjE5NTc2MzAtMmZhZmE4MDAtNWQ4YS00NmRmLThmYjQtOTliNDRiNjRlMjAzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE5VDIzNTMyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFlNGM5OWMyN2UyNTg1OTg2Yjg0ZmMwMzlhNzdkMGEzMzc0ZWMzMWNjNDExZDUxNDUzMjk4ODc1NGQzMzU4MmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.cOARM1mmFPOfuhis_762i7l2qictokEnpaEKT8FlQak)

You can do it actually.

**Comment by patientx** [comment]
> Besides recently we have also added a few cmd flags to tune the performance.
> 
> The below texts are taken from Readme:
> 
> 1. `--always-offload-from-vram` (This flag will make things **slower** but less risky). This option will let Forge always unload models from VRAM. This can be useful if you use multiple software together and want Forge to use less VRAM and give some VRAM to other software, or when you are using some old extensions that will compete vram with Forge, or (very rarely) when you get OOM.
> 2. `--cuda-malloc` (This flag will make things **faster** but more risky). This will ask pytorch to use _cudaMallocAsync_ for tensor malloc. On some profilers I can observe performance gain at millisecond level, but the real speed up on most my devices are often unnoticed (about or less than 0.1 second per image). This cannot be set as default because many users reported issues that the async malloc will crash the program. Users need to enable this cmd flag at their own risk.
> 3. `--cuda-stream` (This flag will make things **faster** but more risky). This will use pytorch CUDA streams (a special type of thread on GPU) to move models and compute tensors simultaneously. This can almost eliminate all model moving time, and speed up SDXL on 30XX/40XX devices with small VRAM (eg, RTX 4050 6GB, RTX 3060 Laptop 6GB, etc) by about 15% to 25%. However, this unfortunately cannot be set as default because I observe higher possibility of pure black images (Nan outputs) on 2060, and higher chance of OOM on 1080 and 2060. When the resolution is large, there is a chance that the computation time of one single attention layer is longer than the time for moving entire model to GPU. When that happens, the next attention layer will OOM since the GPU is filled with the entire model, and no remaining space is available for computing another attention layer. Most overhead detecting methods are not robust enough to be reliable on old devices (in my tests). Users need to enable this cmd flag at their own risk.
> 4. `--pin-shared-memory` (This flag will make things **faster** but more risky). Effective only when used together with `--cuda-stream`. This will offload modules to Shared GPU Memory instead of system RAM when offloading models. On some 30XX/40XX devices with small VRAM (eg, RTX 4050 6GB, RTX 3060 Laptop 6GB, etc), I can observe significant (at least 20%) speed-up for SDXL. However, this unfortunately cannot be set as default because the OOM of Shared GPU Memory is a much more severe problem than common GPU memory OOM. Pytorch does not provide any robust method to unload or detect Shared GPU Memory. Once the Shared GPU Memory OOM, the entire program will crash (observed with SDXL on GTX 1060/1050/1066), and there is no dynamic method to prevent or recover from the crash. Users need to enable this cmd flag at their own risk.

How can we change tile size with this method ? (especially vae tile size) It seems to be using a very small size and with my tests on other gui's I found that I can enter a higher size without getting memory errors andget faster vae decoding. On comfyui with tiled vae 512 and 768 is around the same speed but if I use 960 (which was also the highest I can generate with sd1.5) the gen time speeds up significantly without oom.

**Comment by RedDestiny** [comment]
我发现这个NeverOOM的有个bug，就是经常在图生图模式下报错，而在文生图模式下倒没什么问题：
    Traceback (most recent call last):
      File "D:\SSD\stable-diffusion-webui-forge\modules\call_queue.py", line 57, in f
        res = list(func(*args, **kwargs))
    TypeError: 'NoneType' object is not iterable

我跑的图比较大，主要是vae解码的时候用的显存特别多，用这个可以节省很多显存，加快解码速度。特别需要这个。

**Reply by RedDestiny** [reply]
> 我发现这个NeverOOM的有个bug，就是经常在图生图模式下报错，而在文生图模式下倒没什么问题： Traceback (most recent call last): File "D:\SSD\stable-diffusion-webui-forge\modules\call_queue.py", line 57, in f res = list(func(*args, **kwargs)) TypeError: 'NoneType' object is not iterable
> 
> 我跑的图比较大，主要是vae解码的时候用的显存特别多，用这个可以节省很多显存，加快解码速度。特别需要这个。

我已经知道原因了，是分辨率长宽大小不是SD指定的倍数，每次要比较麻烦地对齐下SD的生成倍数（好像是4还是8的倍数）

**Comment by gigend** [comment]
Does the use of reduced VRAM with NeverOOM affect the speed of generating images?

**Reply by jswag245** [reply]
Yes

**Comment by rm-000** [comment]
I can't get it, where is this Never OOM option in WebUI?

**Reply by SvenErik1968** [reply]
Towards the bottom of the Txt2img and Img2img pages:

![image](https://github.com/user-attachments/assets/b25936fd-0e57-46f7-9c2f-879d94c87287)

**Reply by rm-000** [reply]
that wasn't far :) thanks)

**Comment by rm-000** [comment]
![scr2](https://github.com/user-attachments/assets/e31798b8-c0ca-4bcf-828c-95a6386b5237)
![screenshot](https://github.com/user-attachments/assets/7e3a7594-9ec2-4768-b14b-4fff501dd1f4)
I've tested with different Never OOM options (un)checked, I'm using Flux ([Flux Dev NF4 by lllyasviel](https://github.com/lllyasviel/stable-diffusion-webui-forge/discussions/1079) and [Flux.1-Dev Hyper NF4 + Flux.1-Dev BNB NF4 + Flux.1-Schnell BNB NF4](https://civitai.com/models/638187/flux1-dev-hyper-nf4-flux1-dev-bnb-nf4-flux1-schnell-bnb-nf4)).

Should it really be so that GPU utilization is 0% ? With RTX 4070 1-2 iterations/sec it looks wrong.
Looks like the GPU is not used, although GPU's memory is dedicated
